config/v3_1_experimental/types/disk.go

```diff
@@ -16,72 +16,67 @@ package types
 
 import (
 	"github.com/coreos/ignition/v2/config/shared/errors"
-	"github.com/coreos/ignition/v2/config/validate/report"
+
+	"github.com/ajeddeloh/vcontext/path"
+	"github.com/ajeddeloh/vcontext/report"
 )
 
 func (d Disk) Key() string {
 	return d.Device
 }
 
-func (n Disk) Validate() report.Report {
-	return report.Report{}
-}
-
-func (n Disk) ValidateDevice() (r report.Report) {
+func (n Disk) Validate(c path.ContextPath) (r report.Report) {
 	if len(n.Device) == 0 {
-		r.AddOnError(errors.ErrDiskDeviceRequired)
+		r.AddOnError(c.Append("device"), errors.ErrDiskDeviceRequired)
 		return
 	}
-	r.AddOnError(validatePath(n.Device))
-	return
-}
+	r.AddOnError(c.Append("device"), validatePath(n.Device))
 
-func (n Disk) ValidatePartitions() (r report.Report) {
-	if n.partitionNumbersCollide() {
-		r.AddOnError(errors.ErrPartitionNumbersCollide)
+	if collides, p := n.partitionNumbersCollide(); collides {
+		r.AddOnError(c.Append("partitions", p), errors.ErrPartitionNumbersCollide)
 	}
-	if n.partitionsOverlap() {
-		r.AddOnError(errors.ErrPartitionsOverlap)
+	if overlaps, p := n.partitionsOverlap(); overlaps {
+		r.AddOnError(c.Append("partitions", p), errors.ErrPartitionsOverlap)
 	}
 	if n.partitionsMixZeroesAndNonexistence() {
-		r.AddOnError(errors.ErrZeroesWithShouldNotExist)
+		r.AddOnError(c.Append("partitions"), errors.ErrZeroesWithShouldNotExist)
 	}
-	if n.partitionLabelsCollide() {
-		r.AddOnError(errors.ErrDuplicateLabels)
+	if collides, p := n.partitionLabelsCollide(); collides {
+		r.AddOnError(c.Append("partitions", p), errors.ErrDuplicateLabels)
 	}
 	return
 }
 
 // partitionNumbersCollide returns true if partition numbers in n.Partitions are not unique.
```

Comment:
Minor: could update this docstring (and the following).

Reply:
fixed